sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/BooleanLiteralUnnecessaryBase.cs

```diff
@@ -0,0 +1,286 @@
+        }
+
+        protected static bool ShouldNotReport(ITypeSymbol typeLeft, ITypeSymbol typeRight)
+        {
+            return typeLeft == null
+                || typeRight == null
+                || IsNullableBoolean(typeLeft)
+                || IsNullableBoolean(typeRight);
+        }
+
+        protected static bool IsNullableBoolean(ITypeSymbol type) =>
+            type is INamedTypeSymbol namedType &&
+            namedType.OriginalDefinition.Is(KnownType.System_Nullable_T) &&
+            namedType.TypeArguments.Length == 1 &&
+            namedType.TypeArguments[0].Is(KnownType.System_Boolean);
+
+        protected void CheckForBooleanConstantOnLeftReportOnExtendedLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Extended, context, leftSide: true);
+        }
+
+        protected void CheckForBooleanConstantOnRightReportOnExtendedLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Extended, context, leftSide: false);
+        }
+
+        protected void CheckForBooleanConstantOnLeftReportOnInvertedLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Inverted, context, leftSide: true);
+        }
+
+        protected void CheckForBooleanConstantOnRightReportOnInvertedLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Inverted, context, leftSide: false);
+        }
+
+        protected void CheckForBooleanConstantOnLeftReportOnNormalLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Normal, context, leftSide: true);
+        }
+
+        protected void CheckForBooleanConstantOnRightReportOnNormalLocation(TBinaryExpression binaryExpression,
+            TSyntaxKind booleanSyntaxKind, SyntaxNodeAnalysisContext context)
+        {
+            CheckForBooleanConstant(binaryExpression, booleanSyntaxKind, ErrorLocation.Normal, context, leftSide: false);
+        }
+
+        protected void CheckForBooleanConstant(TBinaryExpression binaryExpression, TSyntaxKind booleanSyntaxKind,
+            ErrorLocation errorLocation, SyntaxNodeAnalysisContext context, bool leftSide)
+        {
+            var expression = leftSide ? Left(binaryExpression) : Right(binaryExpression);
```

Comment:
[cosmetic] We try to split the ternary operators on multiple lines (easier to read).